narwhals/dtypes.py

```diff
@@ -71,10 +73,97 @@ class Object(DType): ...
 class Unknown(DType): ...
 
 
-class Datetime(TemporalType): ...
-
-
-class Duration(TemporalType): ...
+class Datetime(TemporalType):
+    """
+    Data type representing a calendar date and time of day.
+
+    Arguments:
+        time_unit: Unit of time. Defaults to `'us'` (microseconds).
+        time_zone: Time zone string, as defined in zoneinfo (to see valid strings run
+            `import zoneinfo; zoneinfo.available_timezones()` for a full list).
+            When used to match dtypes, can set this to "*" to check for Datetime
+            columns that have any (non-null) timezone.
+
+    Notes:
+        Adapted from Polars implementation at:
+        https://github.com/pola-rs/polars/blob/py-1.7.1/py-polars/polars/datatypes/classes.py#L398-L457
+    """
+
+    def __init__(
+        self: Self,
+        time_unit: Literal["us", "ns", "ms"] = "us",
+        time_zone: str | timezone | None = None,
+    ) -> None:
+        if time_unit not in {"ms", "us", "ns"}:
+            msg = (
+                "invalid `time_unit`"
+                f"\n\nExpected one of {{'ns','us','ms'}}, got {time_unit!r}."
+            )
+            raise ValueError(msg)
+
+        if isinstance(time_zone, timezone):
+            time_zone = str(time_zone)
+
+        self.time_unit = time_unit
+        self.time_zone = time_zone
+
+    def __eq__(self: Self, other: object) -> bool:
+        # allow comparing object instances to class
+        if type(other) is type and issubclass(other, Datetime):
+            return True
+        elif isinstance(other, Datetime):
+            return self.time_unit == other.time_unit and self.time_zone == other.time_zone
+        else:
+            return False
+
+    def __hash__(self: Self) -> int:  # pragma: no cover
+        return hash((self.__class__, self.time_unit, self.time_zone))
```

Comment:
do we need to keep this as `return hash(self.__class__)` so that
```
nw.Datetime('us', 'foo') in {nw.Datetime}
```
keeps working?

It's unfortunate that I'd written the code in Altair like that, but I also don't think it's too bad, it's just a little annoying that it departs from what Polars does

Reply:
Down to only one altair test failing 👀

Reply:
Oh I see, this is due to the change of `__repr__`. The following assertion fails in the altair test:

```py
('Field "timedelta" has type "Duration" which is not supported by Altair'
 in 'Field "timedelta" has type "Duration(time_unit=\'ns\')" which is not supported by Altair. Please ...'
).
```

Reply:
yup - I've opened a PR there to make the test a little laxer https://github.com/vega/altair/pull/3606